Use naive diff for overlay2 when opaque copy up bug present

dmcgowan opened the pull request:
**- What I did**

Update overlay2 driver to use naive diff when kernel has bug which copies up opaque directory flag

**- How I did it**

Added a check on kernels older than 4.8 for the bug.

**- How to verify it**

Run this Dockerfile on kernel without patched fix and should work

```
FROM alpine

RUN mkdir /dir1 && touch /dir1/f1
RUN rm -rf /dir1 && mkdir /dir1 && touch /dir1/f2
RUN touch /dir1/f3 
RUN [ -f /dir1/f2 ]
```

Closes #25244
... for all kernels which support overlay2

Ping @tonistiigi

tonistiigi commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -174,6 +175,14 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 	}
 
+	// If kernel is older than 4.8, check whether opaque copy up bug has been patched
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 8, Minor: 0}) < 0 {
```
We could do this like https://github.com/docker/docker/blob/master/daemon/graphdriver/aufs/aufs.go#L643 with `sync.Once` and lazy loading. Kernel check could probably be skipped then.

dmcgowan replied:
Why should we skip the kernel check though? The fix was merged before 4.8 was released so we can be confident that 4.8 or later have the fixed. Kernel tests were also added around this issue to prevent regression in later versions. Ideally users on newer Kernels won't have to incur the additional startup time required for this check.

tonistiigi replied:
We can keep that check if you like it, although kernel versions are abused all the time and it is hard to predict who will decide to release something with a wrong version. With lazy loading, the additional time for full check should be very small.

tonistiigi reviewed (commented):
Would be good to add a test for #25244

AkihiroSuda commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -174,6 +175,14 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
+	// If kernel is older than 4.8, check whether opaque copy up bug has been patched
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 8, Minor: 0}) < 0 {
+		if err := hasOpaqueCopyUpBug(home); err != nil {
+			logrus.Warnf("Not using native diff for overlay2: %v", err)
```
How about printing something like "Please consider updating the kernel to 4.8.0 or later" here

dmcgowan commented:
@AkihiroSuda added build tag, updated the error to mention updating to fix, and also added to the status output (if this merges first you will need to rebase your other PR).

tonistiigi commented:
@dmcgowan Need to mark that test for linux as well.

dmcgowan committed:
Use naive diff for overlay2 when opaque copy up bug present

When running on a kernel which is not patched for the copy up bug
overlay2 will use the naive diff driver.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

tonistiigi commented:
LGTM

cpuguy83 commented:
LGTM

tristanz commented:
Am I correct to read that this patch means that for kernels without backports yet (RHEL 7), overlay2 will now have same issues with inode exhaustion as overlay in 1.13?

If so, the documentation (https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/) should be updated to reflect that the overlay2 benefits do not apply to kernels older than 4.8 without backports, even though it will work.

tonistiigi commented:
@tristanz No, this only changes the diff algorithm, making `docker commit` bit slower. The optimizations from on disk layout still apply.

dmcgowan commented:
@tristanz the backport does not impact storage for `overlay2`. The only potential impact is running into #21555.

tristanz commented:
@tonistiigi @dmcgowan thanks!   Reading #21555 it looks like this is no longer critical bug after #27209 mitigation.  With this PR and #27209 are any known major issues with overlay2 on RHEL 7.2+ beyond SELinux not working yet?
